lib/blog_web/result.ex

```diff
@@ -0,0 +1,112 @@
+defmodule BlogWeb.OrdGraphQLResult do
+
+  @moduledoc false
+
+  # Produces data fit for external encoding from annotated value tree
+
+  alias Absinthe.{Blueprint, Phase, Type}
+  use Absinthe.Phase
+
+  @spec run(Blueprint.t | Phase.Error.t, Keyword.t) :: {:ok, map}
+  def run(%Blueprint{} = bp, _options \\ []) do
+    result = Map.merge(bp.result, process(bp))
+    {:ok, %{bp | result: result}}
+  end
+
+  defp process(blueprint) do
+    result = case blueprint.execution do
+      %{validation_errors: [], result: result} ->
+        {:ok, data(result, [])}
+      %{validation_errors: errors} ->
+        {:validation_failed, errors}
+    end
+    format_result(result)
+  end
+
+  defp format_result(:execution_failed) do
+    %{data: nil}
+  end
+  defp format_result({:ok, {data, []}}) do
+    %{data: data}
+  end
+  defp format_result({:ok, {data, errors}}) do
+    errors = errors |> Enum.uniq |> Enum.map(&format_error/1)
+    %{data: data, errors: errors}
+  end
+  defp format_result({:validation_failed, errors}) do
+    errors = errors |> Enum.uniq |> Enum.map(&format_error/1)
+    %{errors: errors}
+  end
+  defp format_result({:parse_failed, error}) do
+    %{errors: [format_error(error)]}
+  end
+
+  defp data(%{errors: [_|_] = field_errors}, errors), do: {nil, field_errors ++ errors}
+
+  # Leaf
+  defp data(%{value: nil}, errors), do: {nil, errors}
+  defp data(%{value: value, emitter: emitter}, errors) do
+    value =
+      case Type.unwrap(emitter.schema_node.type) do
+        %Type.Scalar{} = schema_node ->
+          Type.Scalar.serialize(schema_node, value)
+        %Type.Enum{} = schema_node ->
+          Type.Enum.serialize(schema_node, value)
+      end
+    {value, errors}
+  end
+
+  # Object
+  defp data(%{fields: fields}, errors), do: field_data(fields, errors)
+
+  # List
+  defp data(%{values: values}, errors), do: list_data(values, errors)
+
+  defp list_data(fields, errors, acc \\ [])
+  defp list_data([], errors, acc), do: {:lists.reverse(acc), errors}
+  defp list_data([%{errors: errs} = field | fields], errors, acc) do
+    {value, errors} = data(field, errors)
+    list_data(fields, errs ++ errors, [value | acc])
+  end
+
+  defp field_data(fields, errors, acc \\ [])
+  defp field_data([], errors, acc), do: {{:lists.reverse(acc)}, errors}
```

Comment:
This line here in the normal result phase is:

```
defp field_data([], errors, acc), do: {Map.new(acc), errors}
```

This is the only line in the entire file that is different.

Reply:
The main challenge here is that for large results this function gets called a LOT, and so I want to avoid dynamic lookup. One idea is to use macros:

```
defmodule BlogWeb.OrdGraphQLResult do
  use Absinthe.ResultPhase
  
  def to_object_result(acc) do
    {:lists.reverse(acc)}
    # or you could do `OrdMap.new(:lists.reverse(acc))`
  end
end
```